### ISSUE TITLE
Do something useful with this

### ISSUE THREAD

**Issue by freebroccolo** [freebroccolo/purescript-typelevel-arithmetic#1]
@paf31 I don't write a lot of purescript these days so there's not much motivation for me to build this into a really polished and usable library just for my own purposes. However, I think the general approach is pretty solid as far as doing this sort of thing with type classes is concerned. I don't know what your long term plans are for type level stuff in purescript but I would be happy to contribute this development to some larger effort if it is useful.

At the moment, only positive naturals are defined but it is trivial to extend to integers. Rationals are also doable, but require a bit more work. Most major operations are [implemented](https://github.com/freebroccolo/purescript-typelevel-arithmetic/blob/master/src/Main.purs#L58-L78).

There are currently two issues holding this back from being nicer to use in purescript:

1. Using the type level arithmetic in an API is tricky without GADTs.

2. Writing numbers is clunky; builtin/macro support for converting numerals to this representation or at least some sort of macro system that would let us do it in the library.

For (1), I think the only practical answer to this is to just limit the usage to wrapping some existing APIs with a layer that includes type level arithmetic, but which is implemented internally in an unsafe way (e.g., using `unsafeCoerce` on top of functions like `append`). This would almost definitely be better for performance too.

For (2), it's always possible to just precompute a number somewhere and then add it to the source with a comment. It's a bit inconvenient, but maybe that's okay.

The performance of arithmetic operations during compile time is quite good even for some truly massive numbers. Converting `Proxy` to a term level `HugeInt` starts to slow down once you cross a certain threshold but it's still reasonably fast for fairly large numbers—probably larger than will be used much in practice.

The encoding this library uses is that of a tree-based binary which effectively compresses blocks of 0s and 1s. Many of the operations are able to manipulate entire blocks at a time, providing super-exponential performance improvements in the best case. Worst case (i.e., no blocks of length greater than 1) is similar to traditional binary. There are a number of clever ways the worst case

In other words, this implementation favors numbers having low representational complexity with regard to the encoding scheme rather than the usual approach where performance is uniformly determined by the magnitudes of the inputs.

**Comment by paf31** [comment]
@freebroccolo One option could be to move this into `purescript-contrib`. That way, we will keep this up to date as we update the compiler.

There is the new `purescript-typelevel-prelude` library in core too, so I wonder if this could fit there.

**Comment by freebroccolo** [comment]
@paf31 thanks. Moving it to `purescript-contrib` sounds like a good option. I should probably do a little bit of cleanup and add some basic documentation first. I'll try to do that sometime soon.

I'd also be happy to help move things to that prelude lib at some point if you want to try that.
